### PR TITLE
[stdlib] adds fast path to internal func foreignErrorCorrectedGrapheme

### PR DESCRIPTION
**Pull request by valeriyvan**
Adds fast path to internal func foreignErrorCorrectedGrapheme.
Addresses TODO.
Eliminates redundant buffer zeroing in slow path.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by valeriyvan** [comment]
Was always curious how you guys make decision based on all these numbers if fix improves or harms.

**Comment by valeriyvan** [comment]
ping

**Comment by valeriyvan** [comment]
ping

File: stdlib/public/core/UnicodeHelpers.swift [hunk 1]
```diff
@@ -395,17 +395,31 @@ extension _StringGuts {
         startingAt: String.Index(_encodedOffset: start)
       ).0))
     }
+    let range = start..<end
+    let from = self._object.cocoaObject
+
+    let capacity = 16 * 4
```
Comment (milseman) [inline stdlib/public/core/UnicodeHelpers.swift:401]: What is the relationship between this capacity and the actual fixed array size? Why this capacity? Seems like 16 code units would be more than enough for anything non-pathological. What use cases were you seeing?

**Review by milseman — CHANGES_REQUESTED** [review]

File: stdlib/public/core/UnicodeHelpers.swift (hunk 1 above)
Comment (valeriyvan) [reply stdlib/public/core/UnicodeHelpers.swift]: Do you mean this will be enough?

```
// 16 code units would be more than enough for anything non-pathological
_internalInvariant(count <= 16 * MemoryLayout<UInt64>.stride)
var cus = _FixedArray16<UInt64>(allZeros: ())
return cus.withUnsafeMutableBufferPointer { buffer in
  let baseAddress = buffer.baseAddress._unsafelyUnwrappedUnchecked
  let into = UnsafeMutableRawPointer(baseAddress)
    .bindMemory(to: UInt16.self, capacity: 16 * MemoryLayout<UInt64>.stride)
  _cocoaStringCopyCharacters(from: from, range: range, into: into)
  let intoBuffer = UnsafeBufferPointer(start: UnsafePointer<UInt16>(into), count: range.count)
  return Character(String._uncheckedFromUTF16(intoBuffer))
}
```

**Comment by valeriyvan** [comment]
ping

**Comment by valeriyvan** [comment]
pind

**Comment by valeriyvan** [comment]
ping

**Comment by eeckstein** [comment]
@milseman, not sure you saw the "ping"

**Comment by valeriyvan** [comment]
ping

**Comment by valeriyvan** [comment]
ping

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by valeriyvan** [comment]
@milseman?

**Comment by valeriyvan** [comment]
ping

**Comment by valeriyvan** [comment]
ping

**Review by milseman — CHANGES_REQUESTED** [review]
This introduces a significant safety hazard in hard-to-test corner cases for an unimportant optimization. Just remove the `TODO(String performance)` comment, it was added in a world when we thought we'd get small-stack types soon.

**Comment by valeriyvan** [comment]
> This introduces a significant safety hazard in hard-to-test corner cases for an unimportant optimization. Just remove the `TODO(String performance)` comment, it was added in a world when we thought we'd get small-stack types soon.

Do you mean whole PR should become just removing `TODO(String performance)`?

**Comment by milseman** [comment]
Yes

**Comment by valeriyvan** [comment]
PR doesn't make sense anymore
